test/browser-reporters/html.spec.js

```diff
@@ -0,0 +1,1918 @@
+'use strict';
+
+var HTML = require('../../lib/reporters/html');
+
+var Suite = require('../../lib/suite');
+var Context = require('../../lib/context');
+var Test = require('../../lib/test');
+
+describe('HTML reporter', function () {
+  this.slow(250);
+
+  var realOutput;
+  var realOutputStats;
+  var realOutputReport;
+
+  function saveRealOutput () {
+    if (!realOutput) {
+      realOutput = document.getElementById('mocha');
+      realOutputStats = document.getElementById('mocha-stats');
+      realOutputReport = document.getElementById('mocha-report');
+    }
+  }
+
+  function removeRealOutput () {
+    saveRealOutput();
+    if (realOutput) {
+      realOutput.id = null;
+      realOutputStats.id = null;
+      realOutputReport.id = null;
+    }
+  }
+
+  function stubOutput () {
+    removeRealOutput();
+    var tmp = document.createElement('div');
+    tmp.id = 'mocha';
+    document.body.appendChild(tmp);
+    return tmp;
+  }
+
+  function restoreOutput () {
+    var tmp = document.getElementById('mocha');
+    if (tmp && tmp !== realOutput) { tmp.parentNode.removeChild(tmp); }
+    if (realOutput) {
+      realOutput.id = 'mocha';
+      realOutputStats.id = 'mocha-stats';
+      realOutputReport.id = 'mocha-report';
+    }
+    return tmp !== realOutput ? tmp : undefined;
+  }
+
+  before(saveRealOutput); // so that the following afterEach is guaranteed to work
+
+  afterEach(function () { restoreOutput(); }); // WARNING: this is solely a fallback so *further* tests can be guaranteed to report correctly; these tests will not be reported correctly unless `restoreOutput` is called at the end of the test, before throwing any errors
```

Comment:
The comment here might not be true when running in Karma and not opening the debug page -- I don't know enough about how it works. I am certain that it is crucial for using the HTML reporter to report the HTML reporter tests, however (e.g. in a manual page such as the one added in this commit as well, or in Karma's debug usage of Mocha's HTML reporter). I'm considering updating the comment to clarify this. It's the explanation for the try-catching in all the tests, so I hope it's clear enough.